fix(forms): async validator-directives process Observables correctly

choeller opened the pull request:
- **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Bug fix
- **What is the current behavior?** (You can also link to an open issue here)

Currently Async Validator-Directives are not able to return Observables. 

https://github.com/angular/angular/issues/8022
- **What is the new behavior (if this is a feature change)?**

Async Validator-Directives can now return Observables as well as Promises
- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
  
  No
- **Other information**:

while normalizing async-validators the return value of validator-directives was erroneously resolved by `Promise.resolve`. This prevented async-validator-directives to work with Observables correctly

Closes https://github.com/angular/angular/issues/8022

choeller commented:
I agreed to the CLA.

mhevery commented on modules/playground/e2e_test/template_driven_forms/template_driven_forms_spec.ts:
```diff
@@ -18,4 +18,17 @@ describe('Template-Driven Forms', function() {
 
     expect(form.getInnerHtml()).toContain('is invalid credit card number');
   });
+
```
We need a unit test demonstrating the issue. e2e tests should only be used for integration testing. We need to keep the number of e2e tests low otherwise it will take too long to run all of the tests and our velocity will be lowered.

choeller replied:
To be honest I really searched for a while to find unit tests dealing with Template driven forms functionality and didn't find any - so I just decided to extend the existing e2e test. Can you point me to a file to start with? 

The issue is described in https://github.com/angular/angular/issues/8022 So currently it is not possible to return an Observable that does some further chaining (like `map`) in an async validator directive.

choeller replied:
Ok - found the correct testfile in validators_spec.ts. I'm testing that validator-directives work with observables

mhevery commented on modules/@angular/common/src/forms/directives/normalize_validator.ts:
```diff
@@ -11,7 +11,7 @@ export function normalizeValidator(validator: ValidatorFn | Validator): Validato
 export function normalizeAsyncValidator(validator: AsyncValidatorFn | Validator): AsyncValidatorFn {
   if ((<Validator>validator).validate !== undefined) {
-    return (c: AbstractControl) => Promise.resolve((<Validator>validator).validate(c));
+    return (c: AbstractControl) => (<Validator>validator).validate(c);
```
Why is removing the `Promise` the right think to do?

choeller replied:
I'm not entirely sure *why exactly* `Promise.resolve` breaks the Observable chain here, but this function is only supposed to normalize the structure of the returned value - meaning return an `AsyncValidatorFn` that forwards to the `validator.validate` function. The Promise.resolve was introduced in a later commit and since then Template driven forms were not able to use `Obersvables` as return value of a async validator directive

mhevery commented:
Also please rebase onto latest master.

choeller commented:
I'm pretty sure I am on current master - my latest commit is 61b339678dbf61ef4bbcf65ed5d8ca52cea4a139 or am I missing something?

choeller commented:
@mhevery I now added the unit-test to valdiators_spec.ts. Before my change this test failed with

```
Expected Object({ _isScalar: false, source: Observable({ _isScalar: false, _subscribe: Function }), operator: MapOperator({ project: Function, thisArg: undefined }) }) to equal Object({ one: true }).
```

I also found the actual reason for the bug: Because of the Promise.resolve `ObservableMapper.toPromise` was never called. Is the PR ok now? I think having async validation tested in e2e scenario is still worth testing or should I really delete this test again?

mhevery commented:
@vsavkin can you approve this?

vsavkin commented on modules/@angular/common/test/forms/validators_spec.ts:
```diff
@@ -144,6 +161,19 @@ export function main() {
+        ]);
+
+        var value = null;
+        (<Promise<any>>c(new Control("invalid"))).then(v => value = v);
```
why is this type cast required?

vsavkin commented on modules/playground/src/template_driven_forms/index.ts:
```diff
@@ -1,10 +1,12 @@
 import {bootstrap} from '@angular/platform-browser';
-import {Component, Directive, Host} from '@angular/core';
+import {Component, Directive, Host, forwardRef, Provider} from '@angular/core';
+import {Observable} from 'rxjs/Rx'
```
use this instead `import { Observable } from 'rxjs/Observable';`

vsavkin commented:
Some minor comments. LGTM after addressing them.

choeller committed:
fix(forms): async validator-directives process Observables correctly

while normalizing async-validators the return value of validator-directives was erroneously resolved by `Promise.resolve`. This prevented async-validator-directives to work with Observabl

Closes https://github.com/angular/angular/issues/8022

choeller commented:
@kara I'm quite unsure how to rebase this PR. Should I fix the bug in `forms-deprecated` and `forms`? Or would it maybe just make sense that you fix the single line during your refactoring?

choeller commented:
@kara @vsavkin So I now rebased by adding the fix to both - `forms-deprecated` and `forms` and also adressed your comments. Hope this is ok now...

choeller commented:
@kara @vsavkin So I rebased yet another time ;) would be cool to have this either merged or declined, but rebasing it every some days without any reaction is kind of exhausting ;)

vsavkin commented on modules/@angular/common/test/forms-deprecated/validators_spec.ts:
```diff
@@ -14,6 +16,18 @@ export function main() {
+
+    validate(c: any): {[key: string]: any;} {
+      return Observable.create((obs: any) => {
+        let error = this.expected !== c.value ? this.error : null;
```
let => const

vsavkin commented on modules/@angular/common/test/forms-deprecated/validators_spec.ts:
```diff
@@ -80,6 +94,17 @@ export function main() {
     });
 
+    it('should normalize and evaluate async validator-directives correctly', fakeAsync(() => {
+         var c = Validators.composeAsync(
```
var => const

vsavkin commented on modules/@angular/common/test/forms-deprecated/validators_spec.ts:
```diff
@@ -80,6 +94,17 @@ export function main() {
+         var c = Validators.composeAsync(
+             [normalizeAsyncValidator(new AsyncValidatorDirective('expected', {'one': true}))]);
+
+         var value: any = null;
```
var => let

vsavkin commented on modules/@angular/forms/test/validators_spec.ts:
```diff
@@ -14,6 +16,18 @@ export function main() {
+
+    validate(c: any): {[key: string]: any;} {
+      return Observable.create((obs: any) => {
+        let error = this.expected !== c.value ? this.error : null;
```
let => const

vsavkin commented on modules/@angular/forms/test/validators_spec.ts:
```diff
@@ -139,12 +153,22 @@ export function main() {
          }));
 
+      it('should normalize and evaluate async validator-directives correctly', fakeAsync(() => {
+           var c = Validators.composeAsync(
```
var => const

vsavkin commented on modules/@angular/forms/test/validators_spec.ts:
```diff
@@ -139,12 +153,22 @@ export function main() {
+           var c = Validators.composeAsync(
+               [normalizeAsyncValidator(new AsyncValidatorDirective('expected', {'one': true}))]);
+
+           var value: any = null;
```
var => let

vsavkin commented on modules/playground/e2e_test/template_driven_forms/template_driven_forms_spec.ts:
```diff
@@ -18,4 +18,17 @@ describe('Template-Driven Forms', function() {
     expect(form.getInnerHtml()).toContain('is invalid credit card number');
   });
+
+  it('should display async errors', function() {
```
I don't think this test is need. The one in validator_spec is sufficient. If it isn't, adding a test to integration_spec is better. 

Please remove this test.

choeller commented:
@vsavkin Ok - addressed your comments and deleted the e2e-test. Guess you're right that a unit-test ist sufficient here...
